Add Demo Project for RDF4J-Spring

fkleedorfer opened the pull request:
Related Issue: #3387 

Add folder `spring-components/rdf4j-spring-demo` containing a maven project that uses `rdf4j-spring`.

The `rdf4j-spring-demo` project is explicitly **not a maven submodule** of `rdf4j-spring-components` so the pom is self-contained, which makes it clearer to users how to start their own `rdf4j-spring` project.

As discussed in #3387, I am not sure that this is the best place for this demo project, but I think it is the best place I can put it without changes to the rdf4j folder/project structure.

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

jeenbroekstra commented on spring-components/rdf4j-spring-demo/src/main/java/org/eclipse/rdf4j/spring/demo/ArtDemoCli.java:
```diff
@@ -0,0 +1,57 @@
+package org.eclipse.rdf4j.spring.demo;
```
Missing copyright header

jeenbroekstra commented on spring-components/rdf4j-spring-demo/src/main/java/org/eclipse/rdf4j/spring/demo/ArtDemoConfig.java:
```diff
@@ -0,0 +1,28 @@
+package org.eclipse.rdf4j.spring.demo;
```
Missing header

jeenbroekstra commented on spring-components/rdf4j-spring-demo/src/main/java/org/eclipse/rdf4j/spring/demo/dao/ArtistDao.java:
```diff
@@ -0,0 +1,93 @@
+/*
+ * ******************************************************************************
```
This header looks odd - something went wrong with the comment annotation i think? Also noticed this in a couple of other files.

fkleedorfer replied:
It seems this is what happens when you copy/paste the rdf4j header into intellij's license profile. The best I can do with intellij (very helpful especially with new classes) is this:

```
/*
 * *****************************************************************************
 * Copyright (c) 2015 Eclipse RDF4J contributors, Aduna, and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Distribution License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/org/documents/edl-v10.php.
 * *****************************************************************************
 */
```

fkleedorfer replied:
the whole `rdf4j-spring` module has this problem. Is it ok if I fix that in this PR?

jeenbroekstra replied:
I'd really prefer if we could stick to the single, prescribed format for the license header. @hmottestad I know you are an IntelliJ user - can you advise on how you've configured it to use the correctly formatted copyright header? 

@fkleedorfer  as for fixing it in this in PR, that's not a problem.

fkleedorfer replied:
oh, no I meant, I'll make them look the way they are everywhere else, the intellij bit was just meant fyi. It would be nice of course if the formatter could be made to hammer the copyright into the required form, but it's not a problem - it's just a regex replace. So I'll fix that in all the places I touched.

fkleedorfer replied:
done

jeenbroekstra commented on spring-components/rdf4j-spring-demo/src/main/java/org/eclipse/rdf4j/spring/demo/model/EX.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public class EX {
+	private static final String base = "http://example.org/";
+	public static final IRI Artist = SimpleValueFactory.getInstance().createIRI(base, "Artist");
```
Could we replace this with `Values.iri` instead? It's what we're hoping to promote as the recommended pattern these days.

fkleedorfer replied:
I'll look into it. No reservations whatsoever.

jeenbroekstra commented on spring-components/rdf4j-spring-demo/src/main/java/org/eclipse/rdf4j/spring/demo/ArtDemoCli.java:
```diff
@@ -0,0 +1,57 @@
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class ArtDemoCli implements CommandLineRunner {
```
Might be good to put some javadoc on all classes in the demo that briefly explains what it does.

jeenbroekstra reviewed (commented):
Nice work! Particularly enjoying that even after 20+ years, the default example for anything RDF-related is still artists and artworks :)

I'm fine with putting this in this spot for now. We should make an effort to refactor the `examples` module so that we can put this in as an independent submodule before release though.

Left a few inline comments, mostly just about editorial stuff, like copyright headers and javadoc.

fkleedorfer commented:
For the user documentation, I am not sure where to put it - I mean, is it a tutorial, or rather a programming guide, or a tool?. I am going with `site/content/documentation/programming/spring.md` for now.

hmottestad commented on spring-components/rdf4j-spring-demo/pom.xml:
```diff
@@ -0,0 +1,64 @@
+		<dependency>
+			<groupId>org.eclipse.rdf4j</groupId>
+			<artifactId>rdf4j-spring</artifactId>
+			<version>4.0.0-SNAPSHOT</version>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter</artifactId>
+		</dependency>
+	</dependencies>
+	<dependencyManagement>
+		<dependencies>
+			<dependency>
+				<groupId>org.springframework.boot</groupId>
+				<artifactId>spring-boot-dependencies</artifactId>
+				<version>${spring.boot.version}</version>
+				<type>pom</type>
+				<scope>import</scope>
+			</dependency>
+		</dependencies>
+	</dependencyManagement>
+	<build>
+		<plugins>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-compiler-plugin</artifactId>
+				<version>3.8.1</version>
+				<configuration>
+					<fork>true</fork>
+					<release>11</release>
+					<encoding>utf8</encoding>
+				</configuration>
+			</plugin>
+			<plugin>
+				<groupId>org.springframework.boot</groupId>
+				<artifactId>spring-boot-maven-plugin</artifactId>
+				<version>${spring.boot.version}</version>
+			</plugin>
+		</plugins>
+	</build>
```
Do any of these spring-boot plugins or dependencies need to have a CQ logged? As in a works-with CQ. Or maybe it's enough to check clearly defined: https://www.eclipse.org/projects/handbook/#ip-clearlydefined

It's a bit unclear if we need to do anything or not. There is a question about CQ for content on a project website, which is exempt because it's "separate from project content". Dunno if this would be.

> Are project websites subject to the IP Due Diligence Process?
>
> Website content is separate from project content. Project teams are expected to respect licenses for all web site content, but are not required to submit website content (including third-party libraries) for review by the IP Team.

jeenbroekstra replied:
We use spring-boot 2.4.x in other parts of the project, so in that sense this should be fine. I am a little confused why this uses a different patch version though (the actual spring-components module uses 2.4.12).

fkleedorfer replied:
I'll fix that on Monday

fkleedorfer replied:
... I finally decided to change the demo to be a full child module of `rdf4j-spring-components`. It's just less of a hassle.

jeenbroekstra commented:
> For the user documentation, I am not sure where to put it - I mean, is it a tutorial, or rather a programming guide, or a tool? I am going with site/content/documentation/programming/spring.md for now.

That's a good enough place for now. In the end it depends on the writing style (tutorial vs. reference) more than anything else. And we can always reshuffle this if needed.

fkleedorfer commented:
update: the user docs are in the works, I'll probably have a first version finished on Monday

fkleedorfer commented:
I finally decided to make `rdf4j-spring-demo` a submodule of `rdf4j-spring-components` after all. It just makes everything harder not to. Case in point: already now, the spring-boot patch version was different from the one used elsewhere in the project. Also, formatting is so much easier if it's a submodule. So, there's that.

I also fixed the copyright headers in `rdf4j-spring`, and I added the `PredictableUUIDSource` class (and friends).

fkleedorfer commented:
... and the user documentation is finished, so from my point of view, this PR is good to go.

fkleedorfer commented:
Sorry, the unit test support needs more work...

jeenbroekstra commented on spring-components/rdf4j-spring-demo/src/main/java/org/eclipse/rdf4j/spring/demo/ArtDemoCli.java:
```diff
@@ -0,0 +1,90 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Eclipse RDF4J contributors, Aduna, and others.
```
This is not the correct copyright header I'm afraid. It should be this:

```
/*******************************************************************************
 * Copyright (c) 2021 Eclipse RDF4J contributors.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Distribution License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/org/documents/edl-v10.php.
 *******************************************************************************/
```

Note that there's no ", Aduna, and others." <- that is something that is _only_ used for our legacy code from the Sesame project, not for anything new.

See https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting for details.

fkleedorfer replied:
Sorry for that! I just opened a random source file and got the header from there. Not a safe practice, it seems ;-)

fkleedorfer replied:
fixed.

jeenbroekstra commented on spring-components/rdf4j-spring-demo/pom.xml:
```diff
@@ -0,0 +1,75 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ ******************************************************************************
+  ~ Copyright (c) 2015 Eclipse RDF4J contributors, Aduna, and others.
```
No real need to put the header in an xml file, to be honest.

fkleedorfer replied:
ok, removed (from properties file, too)

jeenbroekstra reviewed (commented):
Looks really good! There's unfortunately still a bit of a problem with the copyright header used (see inline comment) but other than that this is good to go.

I've created a followup issue (GH-3428) to look into reorganizing the examples module and moving this demo code there, but for now this is great.
